### ISSUE TITLE
[css-conditional-5][css-nesting-1] Feature detection for nesting

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#8399]
We need to provide a smooth way to transition to nesting, so authors can detect whether nesting is supported by the user agent. So they can provide fallback solutions for UAs that don't support it.

There where already some prior discussions related to this topic, though no explicit issue or resolution regarding this, as far as I can tell.

The major discussion related to this is #8349, but its focus is rather on error recovery than feature detection. Another slightly related one is #2463, though it rather features detection of at-rules and their descriptors.

So I opened this issue to explicitly discuss how nesting support can be made detectable. I believe this is a _must_ for any implementation for shipping nesting capabilities to reach a high level of adoption and avoid page breakages.

Sebastian

**Comment by SebastianZ** [comment]
Maybe the simplest solution would be to reuse the [`selector()` function for that](https://www.w3.org/TR/css-conditional-4/#typedef-supports-selector-fn).

So this could detect whether nesting style rules is supported:

```css
@supports selector(& .foo) {
  …
}
```

Though that only covers style rules. Support for nesting other rules like media queries might require a different approach. One solution would be to support testing of complete rules, as I [suggested earlier](https://github.com/w3c/csswg-drafts/issues/5929#issuecomment-774475760). An example of that could look like this:

```css
@supports (.foo { @media (width >= 500px) { } }) {
  …
}
```

Another, more flexible but also more complex approach would be script based queries, which [I also suggested earlier](https://github.com/w3c/csswg-drafts/issues/3559#issuecomment-1114545949).

Sebastian

**Comment by Loirooriol** [comment]
I highly doubt authors will use something like

```css
@supports selector(& .foo) {
  .wrapper {
    color: magenta;
    & .foo { color: cyan; }
  }
}
@supports not selector(& .foo) {
  .wrapper { color: magenta; } 
  .wrapper .foo { color: cyan }
}
```

They will just use the good ol

```css
.wrapper { color: magenta; } 
.wrapper .foo { color: cyan }
```

The only reasonable fallback strategy seems using JS to read the raw contents of the `<style>`, parse it manually, and if there are nested rules, desugar them. But detecting support for nesting in JS is already possible:

```js
function supportsCSSNesting() {
  try {
    document.querySelector("&");
    return true;
  } catch (e) {
    return false;
  }
}
```

Support for nesting other rules like media queries is also posssible:

```js
var style = document.createElement("style");
style.textContent = ".foo { @media (width >= 500px) { } }";
document.head.append(style);
var {sheet} = style;
style.remove();
return sheet.cssRules[0]?.cssRules?.length === 1;
```

We may want to add nicer ways, but I don't think it's a must.

**Comment by plinss** [comment]
I agree that authors will likely not write single stylesheets with both nested and un-nested versions of selectors. They'll either, as you say, write the stylesheet without nesting until they no longer care about older browsers, or use a preprocessor and let it emit the un-nested version.

However, what's reasonable, and should be encouraged, would be:
```css
@import "nested-and-small.css" supports(selector(&));
@import "not-nested-and-large.css" not supports(selector(&));
```
(or ideally do the equivalent in &lt;link&gt; tags)

This way the author can use a preprocessor to generate the larger, un-nested stylesheet, but the user of a newer browser doesn't have to pay the download cost.

While nesting should be detectable via script, we need to be able to feature detect it without script (or UA sniffing) as well.

**Comment by SebastianZ** [comment]
@plinss I totally agree with what you say and also thought of the usage in `@import` or the HTML variant for `<link>` elements. I guess I should have mentioned that.

@Loirooriol Relying on JS to solve this is also one of my ideas from above, though ideally JS shouldn't be a requirement for feature detection. 

If testing for `selector(&)` is already valid, then we at least have detection support for nesting style rules.

Though being able to test for other nesting features is still important. So there's no gap between implementations and feature detectability.

Sebastian

**Comment by romainmenke** [comment]
```css
@import "nested-and-small.css" supports(selector(&));
@import "not-nested-and-large.css" not supports(selector(&));
```

No browser has implemented this (yet?)

https://wpt.fyi/results/css/css-cascade/import-conditions.html?label=master&label=experimental&aligned&view=subtest&q=import

**Comment by plinss** [comment]
> No browser has implemented this (yet?)

Sure, but no browser has really implemented nesting yet either (STP and Chrome experiments not withstanding). The WG hasn't resolved on the actual mechanism to feature detect nesting yet either, the above was a (reasonable) placeholder. Given there are still open questions on the final syntax of nesting it doesn't make sense to finalize the feature detection yet.

I expect that the WG will finalize how to feature detect nesting when we finalize nesting itself, and any browser implementing nesting will implement the feature detection at the same time (it should be part of a compliant implementation).

**Comment by Loirooriol** [comment]
@plinss In case you are not aware of it, Blink plans to ship the "experiment" in version 112. https://groups.google.com/a/chromium.org/g/blink-dev/c/eFCrkiLynfU/m/JLsh3zQuAAAJ

**Comment by romainmenke** [comment]
> However, what's reasonable, and should be encouraged, would be:
>
> This way the author can use a preprocessor to generate the larger, un-nested stylesheet, but the user of a newer browser doesn't have to pay the download cost.

This is a tangent and I don't want to derail this thread but `@imports ".." not supports(selector(&))` doesn't help browsers without nesting support. Even if `supports` queries for `@import` ship simultaneously with nesting. Any `@imports` with a `supports` condition will be downloaded but never applied.

In theory it is the best method, but because it never shipped it isn't practically useful.

**Comment by plinss** [comment]
> @plinss In case you are not aware of it, Blink plans to ship the "experiment" in version 112.

I'm aware, unfortunately just another example of Google (potentially) being Google and giving the finger to the standards process. (And Apple isn't much better in their messaging around STP).

**Comment by cwilso** [comment]
@plinss I would like to ask you to revisit this comment.  It is absolutely unacceptable to characterize this as Google "giving the finger to the standards process".  We are here, participating in the standards process.  Please do not make ad hominem attacks.

The web standards process is not now, nor has it ever been, a gate that controls features that are shipped in implementations.  At the same time, as a large engine implementer, we have a responsibility to participate in good faith to build consensus on features; I believe we are doing that here.  If you disagree, as per our policy (https://github.com/cwilso/standards-of-behavior), I'm happy to discuss in more depth.

**Comment by plinss** [comment]
@cwilso I edited my comment to remove the phrase that you found offensive. However, while I agree that Google is here, and is participating, I fail to see how shipping a feature in a form that is still under contention and discussion can be characterized as "participating in good faith". You are of course free to ship whatever you want, but you don't get to characterize paving your own path as a "standards process".  Your Blink dev thread makes it clear that your shipping  decision process isn't waiting for this issue to be resolved by the WG, let alone an FO.

**Comment by SebastianZ** [comment]
I guess what we can agree on is that browser vendors are eager to release these nesting capabilities.
So they should also be interested in feature detectability, otherwise the adoption of this big new feature will suffer.

Sebastian

**Comment by SebastianZ** [comment]
@romainmenke Very good point! So it seems that we need a different approach so that UAs _with_ nesting support can block loading stylesheets and rules without nesting to avoid loading them twice in that case.

Sebastian

**Comment by bramus** [comment]
I’ve tested several approaches in the past:

- 1a: [Test for nesting support by actually nesting](https://codepen.io/bramus/pen/oNdrypM)
- 1b: [Test for nesting support by actually nesting + a space toggle](https://codepen.io/bramus/pen/LYmKBEq) for developer convenience.
- 2: [Test for nesting support by testing for `selector(&)` support](https://codepen.io/bramus/pen/XWqLOaR)

The caveat with approach 2 is that it assumes a browser does not ship `&` as a selector without shipping nesting. IIRC @tabatkins mentioned somewhere _(can’t seem to find back where)_ that `&` should not ship separately, so it’s safe to rely on it.

In practice, though, I don’t think authors will test nesting, as they’ll simply see the nested styles not get applied in that case:
- Build a base style layer that doesn’t use nesting
- Layer nested styles on top

**Comment by romainmenke** [comment]
For `@supports selector(&) {}` to be useful it is also critical that browsers ship support for the nesting selector at the root at the same time as actual nesting : https://github.com/w3c/csswg-drafts/issues/5745#issuecomment-1302487789

I've created a pull request with a basic WPT test : https://github.com/web-platform-tests/wpt/pull/38373

Currently Chrome and Safari TP do not support `&` at the root.
If they were to ship today there would be no way to detect support for `&` at the root in the future.

**Comment by yisibl** [comment]
@Loirooriol 

Chrome(e.g. 109) returns `null` for `document.querySelector("&")` in browsers that don't support CSS nesting.

So, the functions detected by JS need a little adjustment:

```js
function supportsCSSNesting() {
  try {
    const s = document.querySelector("&");
    if (!s) return false;
    return true;
  } catch (e) {
    return false;
  }
}
```

Unfortunately, `@supports(selector(&))`  and `CSS.supports('selector(&)')` always returns true.
![image](https://user-images.githubusercontent.com/2784308/217151058-8475eb80-4fba-4671-9d7d-ee4bec19d464.png)

**Comment by sesse** [comment]
> Edit : there was already a test added recently : https://wpt.fyi/results/css/css-nesting/top-level-is-scope.html?label=master&label=experimental&aligned&view=subtest&q=nesting

Even before this, I believe `@supports selector(&)` should have worked? The selector was defined to exist and parse just fine, just not to match anything at the root. (Now it's defined to something else, of course, but that shouldn't affect `@supports`.)

**Comment by romainmenke** [comment]
> Even before this, I believe `@supports selector(&)` should have worked?

It does indeed work, but the question here is if `@supports selector(&)` can be used as proxy for testing actual nesting support.

This approach is only viable if browsers ship nesting and "`&` at the root" at the same time because there is no way to differentiate between them.

**Comment by bramus** [comment]
> It does indeed work, but the question here is if `@supports selector(&)` can be used as proxy for testing actual nesting support.

See [comment above](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1419812180):

>  IIRC @tabatkins mentioned somewhere _(can’t seem to find back where)_ that `&` should not ship separately, so it’s safe to rely on it.

It is the obvious/logical thing to do, but maybe it can also be enforced from within the spec?

**Comment by Loirooriol** [comment]
> I believe `@supports selector(&)` should have worked?

`&` is part of nesting, so the check should have been false with `CSSNesting` disabled.

> It is the obvious/logical thing to do, but maybe it can also be enforced from within the spec?

Seems too late, Blink shipped `&` in https://chromiumdash.appspot.com/commits?commit=336be2dad52bf3454d6063bf36ea4e49a3a74aa7

**Comment by tabatkins** [comment]
Yeah, I *assumed* we wouldn't ship the nesting selector separate from the nesting feature, as it was useless on its own and provided a nice benefit to feature detection. Ah well.

**Comment by sesse** [comment]
Blink shipped & together with nesting support. The commit you're linking to (which did indeed add nesting and & at the same time, but the version that was current at the time) was all behind an experimental flag; it would not do anything in a normal, stable browser.

**Comment by Loirooriol** [comment]
I bisected without experimental flags and I got that commit. And effectively with the stable 109.0.5414.119 snap I can see that `&` is considered a valid selector even if nested rules are disabled.

**Comment by sesse** [comment]
That definitely wasn't the intention of the code, but OK, it's impossible to back and change now.

**Comment by romainmenke** [comment]
> That was fixed in https://bugs.chromium.org/p/chromium/issues/detail?id=1414012 and backport requested to Chrome 111. (Thanks @sesse!)

But impossible to backport to 109 and 110 right?
So `@supports selector (&) {}` is no longer usable as a **reliable** detection?

**Comment by tabatkins** [comment]
Correct, we can't fix existing versions in the wild. (Or rather, we reserve that sort of thing for OMG THE WORLD IS ON FIRE security issues.)

But Chrome versions decay in usage quickly.

**Comment by bramus** [comment]
> So `@supports selector (&) {}` is no longer usable as a **reliable** detection?

_(Cross-posting from Chromium’s I2S)_

Authors can work around Chrome 109+110 incorrectly reporting `&` support by also testing for a feature that ships in any release after Chrome 110. A good candidate would be `cos()`, which ships in Chrome 111: 

```css
@supports selector(&) and (scale: cos(90deg)) {
  …
}
```

This check excludes Chrome 109+110, but includes Chrome 111 with the feature flag on.

More importantly this extended check won’t exclude Safari or Firefox, as they too support `cos()` – from, respectively, versions 15.4 and 108 onwards [according to caniuse.com](https://caniuse.com/mdn-css_types_cos).

**Comment by yisibl** [comment]
To be honest, feature detection here is less likely to affect users migrating to native nesting, instead, I think the more fatal problem is the desugaring to `:is()`.

**Comment by romainmenke** [comment]
> To be honest, feature detection here is less likely to affect users migrating to native nesting, instead, I think the more fatal problem is the desugaring to `:is()`.

This was revisited here : https://github.com/w3c/csswg-drafts/issues/8310
But it isn't related to and doesn't affect feature detection for nesting.

**Comment by LeaVerou** [comment]
> I agree that authors will likely not write single stylesheets with both nested and un-nested versions of selectors. They'll either, as you say, write the stylesheet without nesting until they no longer care about older browsers, or use a preprocessor and let it emit the un-nested version.
> 
> However, what's reasonable, and should be encouraged, would be:
> 
> ```css
> @import "nested-and-small.css" supports(selector(&));
> @import "not-nested-and-large.css" not supports(selector(&));
> ```
> 
> (or ideally do the equivalent in <link> tags)
> 
> This way the author can use a preprocessor to generate the larger, un-nested stylesheet, but the user of a newer browser doesn't have to pay the download cost.
> 
> While nesting should be detectable via script, we need to be able to feature detect it without script (or UA sniffing) as well.

I suspect once Nesting has critical mass, what will be more common would be this pattern:

HTML:
```html
<link rel=stylesheet href="nested.css">
```

nested.css:
```css

@import "compiled-to-nonnested.css" not supports(selector(&));

/* nested CSS */
```

**Comment by romainmenke** [comment]
> `@import "compiled-to-nonnested.css" not supports(selector(&));`

This doesn't work, it was never implemented.
So there is no advantage to it.

**Comment by romainmenke** [comment]
There is also no way to detect support for the even more relaxed syntax where selectors can start with an identifier.

In Chrome 113 :

```js
CSS.supports('selector(& bar)')
// true
CSS.supports('selector(bar &)')
// true
```

Safari TP doesn't seem to work at all :

```js
CSS.supports('selector(&)')
// false
CSS.supports('selector(& bar)')
// false
CSS.supports('selector(bar &)')
// false
```

This seems problematic if the plan is to relax the syntax somewhere in the future.

**Comment by bramus** [comment]
> [#](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1445128344) Safari TP doesn't seem to work at all :
> 
> ```js
> CSS.supports('selector(&)')
> // false
> ```

Hmm, that looks like a bug to me, given Safari TP does support that selector. Searching the WebKit bug tracker, I found this bug filed for it: https://bugs.webkit.org/show_bug.cgi?id=252301

---

> [#](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1445128344) There is also no way to detect support for the even more relaxed syntax where selectors can start with an identifier.

If one could nest inside of `selector()`, it would be possible:

```js
CSS.supports('selector(div { span })');
```

I think this shortcoming should be plugged.

**Comment by romainmenke** [comment]
Chrome is shipping nesting in ±2 weeks.
This issue seems important to give some extra thought before then.

Especially : https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1445128344

> There is also no way to detect support for the even more relaxed syntax where selectors can start with an identifier.

**Comment by SebastianZ** [comment]
> <fantasai> plinss: I believe it's very important that we have a stable and reliable mechanism before we ship anything, can we agree on that
> <fantasai> TabAtkins: I think it's a good idea, I'm not going to commit to it
> <fantasai> plinss: I think shipping this whithout reliable feature detection is a huge mistake

I totally agree with @plinss here. I fear, Google's strategy for shipping nesting without proper support for feature detection will have negative effects. Without a proper mechanism for feature detection, nesting is doomed to either not being adopted or to cause pages to break.

As @romainmenke [pointed out earlier](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1418273286), the main issue is that browsers never implemented `@import "..." supports();`. Without that, authors don't have a global switch for browsers _with_ nesting support and those without. And nesting support can only be checked when wrapping specific rules in `@supports selector(&) { ... }` within a stylesheet.
Unfortunately, using `@supports` in this case mainly means unnecessarily bloating the stylesheet because authors have to provide the same rules twice for browsers with and without support. I.e. they have to wrap every nested rule in `@supports selector(&) { ... }`. And for browsers that don't support nesting they need to add another `@supports not selector(&) { ... }` section basically duplicating their rules.

@bramus

> In practice, though, I don’t think authors will test nesting, as they’ll simply see the nested styles not get applied in that case:
> 
>     * Build a base style layer that doesn’t use nesting
> 
>     * Layer nested styles on top

As outlined above, layering nested styles on top of not nested ones is not really a use case. Authors normally want to apply the same rules independently of how they are written.

Coming back to the issue outlined by @romainmenke, I believe, the main issue is allowing nesting while being backwards compatible _and_ avoid putting the burden onto users.
That means, we need a way for UAs that support nesting to somehow _suppress_ importing external stylesheets. Only when that's possible, nesting can really be used by authors without disadvantages.

Sebastian

**Comment by bramus** [comment]
> [#](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1497825549) The CSS Working Group just discussed `[css-conditional-5][css-nesting-1] Feature detection for nesting`.
> …
> < heycam> @supports rule(p { & { } })
> < fremy> @supports (&{})
> < bradk> @supports (nesting) { … }

I like @heycam’s suggestion here to be able to pass an entire rule-set into `@supports`. Of the three suggestions, it has the widest coverage.

This would help not only here with nesting, but also cover future/other CSS features such as `@scope`, as authors can pass in `@scope (.foo) to (.bar) { & { color: hotpink; } }` to detect if that specific syntax is supported or not. Using a keyword like `nesting` doesn’t cover that.

This extension to `@supports` would pair really nicely with [`@import … supports(…)`](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1416907375) _(suggested by @plinss above)_ and [`link[media="supports()"]`](https://github.com/whatwg/html/issues/7540) _(something that was discussed for Cascade Layers)_.

This extension to `@supports` doesn’t seem mandatory to be able to detect nesting though, as passing `selector(&)` will cover _“is nesting supported?”_ just fine as the reality is that nesting and `&` ship together. Yes, some pre-release versions of Chrome/Safari incorrectly claim (or don’t claim) support for `&`, but given that these are pre-releases or versions that require the user to enable a feature flag, I think it’s safe to assume that users of those versions update their browsers very often.

**Comment by romainmenke** [comment]
> Yes, some pre-release versions of Chrome/Safari incorrectly claim (or don’t claim) support for &, but given that these are pre-releases or versions that require the user to enable a feature flag, I think it’s safe to assume that users of those versions update their browsers very often.

It wasn't a pre-release version of Chrome that incorrectly claimed to support `&` :)

There also isn't a way to differentiate between what shipped in Chrome 112 and the highly likely direction of the specification :

```js
// In Chrome 112
CSS.supports('selector(&)')
// true
CSS.supports('selector(& div)')
// true
CSS.supports('selector(div &)')
// true
```

Ideally authors have a reliable way to detect support for the final version of nesting, even in older versions of Chrome.

**Comment by bramus** [comment]
> It wasn't a pre-release version of Chrome that incorrectly claimed to support `&` :)
True. There’s [a workaround](https://github.com/w3c/csswg-drafts/issues/8399#issuecomment-1427847117) – which covers all browsers – I mentioned before, but admittedly it’s a nasty hack.

> There also isn't a way to differentiate between what shipped in Chrome 112 and the highly likely direction of the specification

If more robust/reliable feature detection through _< method-this-issue-settles-on >_ lands, Chrome 112 wouldn’t support that new detection method, so it would still get the non-nested styles because it has a more limited nesting support.

---

Rewinding a bit. Say the suggested `@import … supports(…);` lands. Currently using those statements in any browser does not load any of the referenced CSS files, as they don’t understand the statements.

```css
@import "nested-and-small.css" supports(<method-to-detect-nesting>);
@import "not-nested-and-large.css" not supports(<method-to-detect-nesting>);
```

That means, the snippet above won’t cut things here:
- Browsers that support `@import … supports(…);` + nesting will load `nested-and-small.css`.
- Browsers that don’t support `@import … supports(…);` will load none of those files.

What is parseable is, is this … but then you’d end up with double loads + competing stylesheets in browsers that do support `@import … supports(…);`

```css
@import "not-nested-and-large.css";
@import "nested-and-small.css" supports(<method-to-detect-nesting>);
```

Not sure how this could be solved in a nice way 🤔

**Comment by romainmenke** [comment]
Correct, although browsers do download all stylesheets, even when they ignore them.
Bandwidth usage is roughly doubled for end users and they still have a broken page.

------

Some browsers maybe don't intend to ship nesting support in the sort term.
These could add support for `@import … supports(…);`. This makes it possible to have a transition period where some browsers have nesting support "now" and others only "later".

It doesn't help all users of older browser versions but it might still be a useful tool.

**Comment by SebastianZ** [comment]
Summarizing the remaining parts:

1. Feature detection of nesting beyond style rules
2. Avoid double-loading stylesheets for browsers that do support nesting and feature detection

The first poing still directly refers to this issue. For the second point I can create a separate issue.

Sebastian

**Comment by yisibl** [comment]
> There is also no way to detect support for the even more relaxed syntax where selectors can start with an identifier.
> 
> In Chrome 113 :
> 
> CSS.supports('selector(& bar)')
> // true
> CSS.supports('selector(bar &)')
> // true
> Safari TP doesn't seem to work at all :
> 
> CSS.supports('selector(&)')
> // false
> CSS.supports('selector(& bar)')
> // false
> CSS.supports('selector(bar &)')
> // false
> This seems problematic if the plan is to relax the syntax somewhere in the future.

## This is a memo:

Safari [bug 252301](https://bugs.webkit.org/show_bug.cgi?id=252301 ) was fixed in [Safari TP 166](https://webkit.org/blog/13964/release-notes-for-safari-technology-preview-166/#:~:text=Implemented%20%40supports%20with%20the%20%26%20nesting%20operator). 

| Version | Release Date | Relevant Changes |
|---|---|---|
| [Safari TP 166](https://webkit.org/blog/13964/release-notes-for-safari-technology-preview-166/) | 2023-03-23 | Implemented `@supports` with the `&` nesting operator. |
| [Safari TP 167](https://webkit.org/blog/14100/release-notes-for-safari-technology-preview-167/) | 2023-04-05 | Added support for the top-level `&` selector for CSS Nesting. |
| [Safari 16.5](https://webkit.org/blog/14154/webkit-features-in-safari-16-5/) | 2023-05-18 | Shipped CSS Nesting support in a stable release. |

I also tested it in Safari 16.6 and confirmed that it works.

<img width="500"  alt="Image" src="https://github.com/user-attachments/assets/74a62b28-3cb6-4da9-ab8c-62bffa71869f" />
